Process skipped clients

junaidwarsivd opened the pull request:
issue ref :  https://github.com/opensrp/support/issues/267

ekigamba committed:
Append APK file name with version and variant

ekigamba committed:
Separate oauth client id and secret for different variants

ekigamba committed:
Comment out code in client processor

junaidwarsivd committed:
added broadcast reciever and service to reprocess the client

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/job/GizReProcessJob.java:
```diff
@@ -0,0 +1,27 @@
+import org.smartregister.giz.service.ReProcessSyncIntentService;
+import org.smartregister.job.BaseJob;
+
+public class GizReProcessJob extends BaseJob {
```
Rename this to `GIZClientReprocessJob` for now

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/recievers/SyncStatusReciever.java:
```diff
@@ -0,0 +1,40 @@
+            if (fetchStatusSerializable instanceof FetchStatus) {
+                FetchStatus fetchStatus = (FetchStatus) fetchStatusSerializable;
+                    if (fetchStatus.equals(FetchStatus.fetched)) {
+                        GizReProcessJob.scheduleJobImmediately(GizReProcessJob.TAG);
```
I believe we want to call the job once sync is complete eg sync is done fetching all the latest records and not successfully fetched one batch. I'm not sure if this is what is currently implemented

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+
+
+        try {
+            if(formSubmissionIDs.size()>0)
```
Reformat the code to fix spacing in the if statement

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+            if(formSubmissionIDs.size()>0)
+            GizUtils.initiateEventProcessing(formSubmissionIDs);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
```
Use Timber to print exceptions. It sends exceptions to Crashlytics

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+        try {
+            if(formSubmissionIDs.size()>0)
+            GizUtils.initiateEventProcessing(formSubmissionIDs);
+        } catch (Exception e) {
```
I don't think we should be generally catching Exception here

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+    {
+        String missingClientRegister = "SELECT DISTINCT baseEntityId, formSubmissionId AS formSubmissionId FROM event WHERE baseEntityId IN (SELECT baseEntityId FROM client WHERE baseEntityId NOT IN (SELECT DISTINCT base_entity_id FROM client_register_type)) AND eventType LIKE '%Registration%'";
+        String missingECClient = "SELECT DISTINCT baseEntityId, formSubmissionId AS formSubmissionId FROM event WHERE baseEntityId IN (SELECT baseEntityId FROM client WHERE baseEntityId NOT IN (SELECT DISTINCT base_entity_id FROM ec_client)) AND eventType LIKE '%Registration%'";
+        List<String> formSubmissionIDs = new ArrayList<>();
```
This list allows duplicates which might not be handled

ekigamba commented:
Fix codacy and build checks. Kindly make sure you fix this before requesting for a review

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/recievers/SyncStatusReciever.java:
```diff
@@ -0,0 +1,40 @@
+
+import static org.smartregister.receiver.SyncStatusBroadcastReceiver.EXTRA_FETCH_STATUS;
+
+public class SyncStatusReciever  extends BroadcastReceiver {
```
Rename this to `GizReprocessSyncStatusReceiver` since this is only being used by the `GizClientReprocessJob`

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+        SQLiteDatabase database = getReadableDatabase();
+        Cursor cursor = database.rawQuery(query, null);
+        int columnIndex = cursor.getColumnIndex("formSubmissionId");
+        if(cursor.getCount() >  0) {
```
Reformat the code file to fix the spacing

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +25,37 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
         return hasEvent;
     }
+
+    public void ReprocessClients()
```
:smile: `camelCase` please and let's rename this to `processSkippedClients` or something a bit more descriptive enough that only unprocessed clients are being processed here

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/service/ReProcessSyncIntentService.java:
```diff
@@ -0,0 +1,27 @@
+
+public class ReProcessSyncIntentService extends IntentService {
+
+    public static final String TAG = "ReValidateSyncIntentService";
```
Let's have the tag similar to the name of the service

junaidwarsivd committed:
added the corrections in the PR review

junaidwarsivd committed:
added UnitTest for GizEventRepository

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/job/GIZClientReprocessJob.java:
```diff
@@ -0,0 +1,27 @@
+
+public class GIZClientReprocessJob extends BaseJob {
+
+    public static final String TAG = "GizReValidationJob";
```
Change this tag also to match the Job

junaidwarsivd committed:
added more corrections to the PR changed the job trigger to isComplete action
added a check for existing formsubmission id while population the list
corrected the job tag

junaidwarsivd committed:
corrected the condition

junaidwarsivd committed:
removed unused imports

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/recievers/GizReprocessSyncStatusReceiver.java:
```diff
@@ -0,0 +1,27 @@
+            if(isComplete)
+            {
+                GIZClientReprocessJob.scheduleJobImmediately(GIZClientReprocessJob.TAG);
+                Timber.d("fetch completed");
```
This log might be misleading

junaidwarsivd committed:
reformatted the code

ekigamba commented on opensrp-giz-malawi/build.gradle:
```diff
@@ -404,6 +404,8 @@ dependencies {
 
     implementation "androidx.multidex:multidex:2.0.1"
     testImplementation "androidx.work:work-testing:2.7.1"
+    androidTestImplementation 'androidx.test:runner:1.3.0'
```
These two are also applied for instrumentation tests that are not changed in this PR. Confirm that we need these and probably change this to `testImplementation`

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +27,34 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+        if (cursor.getCount() > 0) {
+            cursor.moveToFirst();
+            do {
+                if (!ids.contains(cursor.getString(columnIndex)))
```
Time complexity for contains is n. I still believe that a HashSet is better here

junaidwarsivd committed:
updated according to the PR feedback

junaidwarsivd committed:
changed androidTestImplemenation to testImplementation

junaidwarsivd committed:
codacy issues removed

hilpitome commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +27,33 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
     }
+
+    public void processSkippedClients() {
+        String missingClientRegister = "SELECT DISTINCT baseEntityId, formSubmissionId AS formSubmissionId FROM event WHERE baseEntityId IN (SELECT baseEntityId FROM client WHERE baseEntityId NOT IN (SELECT DISTINCT base_entity_id FROM client_register_type)) AND eventType LIKE '%Registration%'";
```
I am not sure we need to alias formSubmissionId here

hilpitome commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/GizEventRepository.java:
```diff
@@ -21,4 +27,33 @@ public boolean hasEvent(@NonNull String baseEntityId, @NonNull String eventType)
+
+    public void processSkippedClients() {
+        String missingClientRegister = "SELECT DISTINCT baseEntityId, formSubmissionId AS formSubmissionId FROM event WHERE baseEntityId IN (SELECT baseEntityId FROM client WHERE baseEntityId NOT IN (SELECT DISTINCT base_entity_id FROM client_register_type)) AND eventType LIKE '%Registration%'";
+        String missingECClient = "SELECT DISTINCT baseEntityId, formSubmissionId AS formSubmissionId FROM event WHERE baseEntityId IN (SELECT baseEntityId FROM client WHERE baseEntityId NOT IN (SELECT DISTINCT base_entity_id FROM ec_client)) AND eventType LIKE '%Registration%'";
```
check previous comment on formSubmissionId

junaidwarsivd committed:
added android 12 compatibility

ekigamba commented on opensrp-giz-malawi/src/main/AndroidManifest.xml:
```diff
@@ -280,18 +281,18 @@
-        <service android:name="org.smartregister.sync.intent.SettingsSyncIntentService" />
-        <service android:name="org.smartregister.sync.intent.SyncIntentService" />
-        <service android:name=".service.ReProcessSyncIntentService"/>
+        <service android:name="org.smartregister.immunization.service.intent.VaccineIntentService" android:exported="true" />
```
Why are these services set as `exported` and why is this explicitly declared?

ekigamba commented on opensrp-giz-malawi/src/main/java/org/smartregister/giz/recievers/GizReprocessSyncStatusReceiver.java:
```diff
@@ -0,0 +1,28 @@
+            }
+        }
+
+        Timber.d("broadcast Recieved");
```
I think the log is too vague and it would be easier to understand if it was a bit more specific on the broadcast that was received

ekigamba committed:
Code cleanup

ekigamba committed:
Log skipped clients for reprocessing

ekigamba committed:
Process skipped clients events in-order of eventDate after processing clients

- Update versionCode and versionName to 47 and 0.4.7

ekigamba committed:
Fix commented out code in client processor

ekigamba committed:
Update ANC version to fix next button in form

- Update immunization library to fix duplicate vaccines and services
- Fix client-core version

ekigamba committed:
Fix OPD Close event processing

- Add config to close ec_client records when an OPD Close event is encountered
- Filter out closed records from the OPD register

ekigamba committed:
Mark deceased children as closed in ec_client

- Update child register query to use ec_client.is_removed flag

ekigamba committed:
Update README setup instructions

ekigamba committed:
Remove unused imports

ekigamba committed:
Fix PNC registration form cannot move to the next page

Fixes https://github.com/opensrp/opensrp-client-giz-malawi/issues/510

ekigamba committed:
Process death and OPD close events

- These events had not been previously processed due to a bug

ekigamba committed:
Fix All Clients register search to remove dead clients

ekigamba committed:
Code cleanup

ekigamba committed:
Update versionName and versionCode to 0.4.8 and 48

ekigamba committed:
Fix Child register search showing deceased clients

ekigamba committed:
Update versionName and versionCode to 0.4.9 and 49
